### PR TITLE
Feature/647 value dist reducer non pairwise

### LINKED ISSUE #647: AnalyzerResultReducer for Value distribution results

### PR DESCRIPTION
**Pull request by tomaszguzialek**
This PR is competitive to #660, so when reviewing have a look at both approaches and merge only one. This approach is iterative reducing while #660 merges one partial result at a time.

Fixes #647

### CODE REVIEW

File: components/value-distribution/src/main/java/org/datacleaner/result/ValueCountingAnalyzerResult.java [hunk 1]
```diff
@@ -61,7 +61,7 @@
 
     public Integer getUnexpectedValueCount();
 
-    public boolean hasAnnotatedRows(String value);
+    public Boolean hasAnnotatedRows(String value);
```
Comment (kaspersorensen) [inline components/value-distribution/src/main/java/org/datacleaner/result/ValueCountingAnalyzerResult.java]: Why the change? Do you ever want to allow a null here?

File: engine/env/spark/src/main/java/org/datacleaner/spark/SparkAnalysisRunner.java [hunk 2]
```diff
@@ -54,9 +54,23 @@
...
+        this(sparkContext, sparkJobContext, null);
+    }
+
+    public SparkAnalysisRunner(JavaSparkContext sparkContext, SparkJobContext sparkJobContext, Integer minPartitions) {
         _sparkContext = sparkContext;
         _sparkJobContext = sparkJobContext;
+        if ((minPartitions != null) && (minPartitions > 0)) {
+            _minPartitions = minPartitions;
+        } else {
+            logger.warn(
```
Comment (kaspersorensen) [inline engine/env/spark/src/main/java/org/datacleaner/spark/SparkAnalysisRunner.java]: Here it's going to raise a warning for something that is the default behaviour ... I would rather only throw an exception if minPartitions is non-null and less-or-equal to 0.

File: components/value-distribution/src/main/java/org/datacleaner/result/ValueCountingAnalyzerResult.java (hunk 1 above)
Comment (tomaszguzialek) [inline components/value-distribution/src/main/java/org/datacleaner/result/ValueCountingAnalyzerResult.java]: Yes, returning null if GroupedValueDistribution has more than one child result.

File: engine/env/spark/src/main/java/org/datacleaner/spark/SparkAnalysisRunner.java (hunk 2 above)
Comment (tomaszguzialek) [inline engine/env/spark/src/main/java/org/datacleaner/spark/SparkAnalysisRunner.java]: True. The initial plan was to warn only if minPartitions parameter was specified, but an incorrect value. Will fix it.

File: components/value-distribution/src/main/java/org/datacleaner/result/ValueCountingAnalyzerResult.java (hunk 1 above)
Comment (kaspersorensen) [inline components/value-distribution/src/main/java/org/datacleaner/result/ValueCountingAnalyzerResult.java]: Hmm that's very subtle ... Hard to guess even. I would rather just return false then and require the code to call this on the child results.

Comment (kaspersorensen) [inline components/value-distribution/src/main/java/org/datacleaner/result/ValueCountingAnalyzerResult.java]: Right now any calls like this will start throwing exceptions:

```
if (result.hasAnnotatedRows("foo")) {
 ...
}
```

We might better then rename the method so that we can at least catch it at compile time. But in general I don't care about trying to patch the interface here - it's a beast that eventually needs to be redone, but a lot of consideration needs to go into it because at the same time the analyzer has to return some thing from which metrics can automatically be gathered.

File: components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        Collection<ValueFrequency> flattenedTopValues = getOrInitializeFlattenedTopValues(flattenedTopValuesMap,
+                singlePartialResult);
+        Collection<String> flattenedUniqueValues = getOrInitializeFlattenedUniqueValues(flattenedUniqueValuesMap,
+                singlePartialResult);
+        Collection<ValueFrequency> flattenedDistinctValues = getOrInitializeFlattenedTopValues(
+                flattenedDistinctValuesMap, singlePartialResult);
+
+        flattenedUniqueValues.addAll(singlePartialResult.getUniqueValues());
+        flattenedTopValues.addAll(singlePartialResult.getTopValues().getValueCounts());
+        flattenedDistinctValues.addAll(singlePartialResult.getValueCounts());
```
Comment (kaspersorensen) [inline components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java]: What I am noticing is that ```SingleValueDistributionResult.getValueCounts()``` actually contains all the data you need here ... both all the "top values", the nulls and the uniques. Would it maybe be easier to just get that list of value counts and make up all the rest from it? That way we could reduce the complexity here I think and also avoid moving the same data around into multiple collections.

**Comment by kaspersorensen** [comment]
I prefer this approach over #660 ... shall we close the other one?

**Comment by tomaszguzialek** [comment]
Introducing ReducedValueDistributionResult simplified the code, but it solves the problem only for SingleValueDistributionResults. It does not cover GroupedValueDistributionResults...

**Comment by kaspersorensen** [comment]
Not having time to look at the code, I could still imagine that a grouped result simply holds a list of underlying results. Now the type of the underlying results would be ReducedValueDistributionResult instead of the old SingleValueDistributionResult. No?

**Comment by tomaszguzialek** [comment]
Internally, I could hold a collection of child results, but I need to expose them as collection to the UI. The interface currently does not have any method for this...

**Comment by kaspersorensen** [comment]
I'm not sure we are talking about the same thing then? When I look at GroupedValueDistributionResult I see this method:
```
@Override
    public Collection<? extends ValueCountingAnalyzerResult> getGroupResults() { ... }
```
Since that's just "anything that extends ```ValueCountingAnalyzerResult``` that could also be the reduced ones?

**Comment by tomaszguzialek** [comment]
OK, I haven't seen this. It is a different interface implementing this method. However, I feel we are back in the same place... I can internally hold ```Map<String, Map<String, Integer>>``` representing ```group -> value -> count```. But what should I do with methods like ```public Collection<String> getUniqueValues()?``` Flatten the map (some locally unique values will be then removed)?

**Comment by kaspersorensen** [comment]
hmm you shouldn't re-implement the grouped one I think. It is already pretty OK. You just need some awareness of groups when you combine all the single results.

**Comment by kaspersorensen** [comment]
But yes, in the reducer I think you will need to temporarily keep a ```Map<String, Map<String,Integer>>``` in order to keep the state before building first a ```Map<String, ReducedValueDistributionResult>``` and then turn that into a single ```GroupedValueDistributionResult```.

**Comment by tomaszguzialek** [comment]
Ready for review again...

**Comment by kaspersorensen** [comment]
LGTM! I will manually apply this after squashing it to avoid a confusing commit log :)
